pkgs/build-support/trivial-builders/test-overriding.nix

```
@@ -0,0 +1,75 @@
+# Invoked from pkgs/test/default.nix
+{ lib
+, runCommand
+, callPackage
+, writeShellScript
+}:
+let
+  extglobScript = ''
+    shopt -s extglob
+    touch success
+    echo @(success|failure)
+    rm success
+  '';
+
+  # Check for extglobScript, so we're assuming it's writeShellScript's
+  # checkPhase which we reuse, but enable extglob beforehand.
+  allowExtglob = old: {
+    checkPhase = ''
+      export BASHOPTS
+      shopt -s extglob
+      ${old.checkPhase}
+      # Technically, if you want to restore the previous state, you want to save
+      # the previous state and restore it, instead of set/unset + export/unexport;
+      # this is just meant as a simple example.
+      shopt -u extglob
+      export -n BASHOPTS
+    '';
+  };
+
+  # Run old checkPhase, but only succeed if it fails
+  assertFail = old: {
+    # write old checkPhase into a shell script, so we can check for
+    # the phase to fail even though we have `set -e`.
+    checkPhase = ''
+      if source "${writeShellScript "old-check-phase" old.checkPhase}" 2>/dev/null; then
```

```suggestion
      if source ${writeShellScript "old-check-phase" old.checkPhase} 2>/dev/null; then
```
Correct escaping would require `lib.escapeShellArg`, but the convention for interpolated store paths seems to be to leave them unquoted.

---

I personally think it's a bad convention - but yes then use escapeShellArg.

---

store paths themselves are guaranteed to neither need escaping nor quoting, but not their contents. So in this case it would be safe to do neither indeed.

---

Wait, really? Is that documented somewhere?

---

Nope, counterexample:

```nix
pkgs.writeText "?" ""
```

produces a store path with a `?`

---

And in the general case (though not relevant here): I would expect file names in a store directory not to have a restriction.
(We were discussing a test for writeShellScriptBin, where it would be relevant - except for that the name would be hardcoded in the test.)

---

Indeed.

---

(Also, I think the Nix store prefix itself can be anything.)

---

There are some restrictions (i. e. `lib.sanitizeDerivationName`), but I guess in this case we don't handle arbitrary derivation names. I'm usually in favor of using `escapShellArg`, but here it'd just clutter the code unnecessarily and make it less readable, imo.

---

I agree.